enhancements/catalog-source-config.md

```diff
@@ -0,0 +1,340 @@
+`spec.configMap` gets used, but `spec.address` and `spec.image` get ignored. Since there could be attributes defined that just get ignored by the controller, it could be hard for a user to know what is happening given just a `CatalogSource` CR and no documentation.
+
+### Options
+
+Three different options are outlined below. The first two would require breaking changes to the API, while the third would be the "minimal" optional. The path taken likely depends on the whether we are planning for a new `CatalogSource` CRD version in the near future or not.
+
+#### Union Types
+
+We use a union type to define the type and configuration for the `CatalogSource`. This can be realised with the [`oneOf`](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/) keyword supported in OpenAPI v3 CRD schemas. The schema would then look something like:
+
+```yaml
+spec:
+  type: object
+  required:
+    - sourceConfig
+  properties:
+    # Deprecate 
+    # image: ... 
+    # address: ... 
+    # configMap: ...
+    # sourceType: grpc | configMap | internal
+    
+    # Keep as is:
+    description: ...
+    displayName: ...
+    icon: ...
+    priority: ...
+    publisher: ...
+    secrets: ...
+
+    # Add:
+    # Note: 'internal' type is being deprecated so it gets no config structure
+    grpcPodSource:
+      podSpec:  # a subset (or the whole of) corev1.PodSpec
+        image: ... # required
+        # all other attributes are optional
+        # and override the default catalog source pod spec
+        ...
+      ...
+    grpcServiceSource:
+      address: ...
+      ...
+    configMapSource:
+      configMap: ...
+      ... 
+  oneOf:
+    - required: ["grpcPodSource"]
+    - required: ["grpcServiceSource"]
+    - required: ["configMapSource"]
+                  
+``` 
+
+This option introduces breaking changes. This means that while we are on the depracation path, the users will be able to configure their `CatalogSource`s in the current way, _OR_, use one of `grpcPodSource`, `grpcServiceSource`, or `configMapSource` (with precendence given to the new configuration). 
+
+**Pros**
+  1. Each CatalogSource type has its own configuration structure
+  2. We don't polute the top level with endless optional attributes that behave differently depending on the values of other attributes
+  3. Better UX
+  4. New catalog source types can be added easily
+
+
+**Cons**
+  1. Introduces breaking changes
+  2. There is not first-class support for union type validation in kubebuilder/operator-sdk. We'll need another mechanism to add and test these validations (e.g. patch that gets executed after `make controller-gen` in the `api` repo)
+
```

Comment:
the bundle spec is validated when we run `make bundle`
All checks and specifications are defined in https://github.com/operator-framework/api. So, why would we need a new mechanism? Why the changes in the spec here cannot be checked via current checks to validate the bundle spec?

Reply:
I don't think I expressed myself properly here. What I mean is there are currently no annotations given by kubebuilder that can add the `oneOf` to the schema. So, we'll need to build our own automation around this.

Reply:
Might be good to specifically call out that what you're talking about here is OpenAPIv3 validation via the CRD schema, such that the apiserver can validate correct combinations of fields on our behalf.

Reply:
I'll update this in the next iteration